File: src/javatruco/Deck.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package javatruco;

/**
 *
 * @author kaique
 */
public class Deck implements PuxaCarta{
    private Carta[] deck;
    private int topo;
    public static final int NUM_CARTAS = 40;
    
    public Deck(){
        deck = new Carta[NUM_CARTAS];
        fazerMonte();
    }
    
    public void fazerMonte(){
        criaDeck();
        embaralha();
    }
    
    private void criaDeck(){
        topo = NUM_CARTAS-1;
        for(int v=0; v<10; v++){
            for(int n=0; n<4; n++){
                deck[v*4+n] = new Carta(v, n);
            }
        }
    }
    
    private int sorteia(int n){
        return (int)(n*Math.random());
    }
    
    private void embaralha(){
        Carta aux;
        int sort;
        
        for(int i=0; i<NUM_CARTAS;i++){
            sort = sorteia(NUM_CARTAS-i);
            aux = deck[sort];
            deck[sort] = deck[NUM_CARTAS-i-1];
            deck[NUM_CARTAS-i-1] = aux;
        }
    }
    
    public void imprimeDeck(){
        for(int i=0; i<NUM_CARTAS; i++){
            System.out.print(deck[i].Desenho());
        }
        System.out.print("\n");
    }

    @Override
    public Carta obterCarta() {
        Carta c = deck[topo];
        topo--;
        return c;
    }
}

File: src/ctruco/Deck.c
/**************
 * Constantes *
 **************/
#define _NUM_CARTAS 40
/*************
 * Atributos *
 *************/
var(public, Object*, deck);
var(public, int, topo);
/**********
 * Classe *
 **********/
class(Deck, {&deck, &topo});// <-- Deve implementar a interface "PuxaCarta"
/**************
 * Construtor *
 **************/
constructor(Deck, 
	get(deck) = malloc(sizeof(int)*_NUM_CARTAS);
	call(&this, fazerMonte);
	);
/***********
 * Funções *
 ***********/
/*
 * Deck:fazerMonte - 
 * Cria o monte e embaralha.
 */
function(Deck, public, fazerMonte,
	call(&this, criaDeck);
	call(&this, embaralha);
	);
/*
 * Deck:criaDeck - 
 * Cria o monte de cartas em ordem crescente de valor
 */
function(Deck, public, criaDeck,
	int v, n;
	//puts("<< GRANDE POTENCIAL DE ERRO 1 >>");
	get(topo) = (generic)(intptr_t)_NUM_CARTAS -1;
	for(v=0;v<10;v++){
		for(n=0; n<4; n++){
			instanciate(Carta, nova_carta, v, n);
			((Object*)get(deck))[v*4+n] = *nova_carta;
			//printf("Carta: %s\n", (char*)call(&((Object*)get(deck))[v*4+n], Desenho));
		}
	}
	);
/*
 * Deck:sorteia - 
 * Recebe um int n, e retorna um generic int sortiado entre 0 e n.
 */
function(Deck, public, sorteia, 
	arg(int, n);
	return (generic)(intptr_t)(int)(rand()%n);
	);
/*
 * Deck:embaralha - 
 * Embaralha o monte.
 */
function(Deck, public, embaralha,
	Object aux;
	int sort, i;
	//puts("<< GRANDE POTENCIAL DE ERRO 2 >>");
	for(i=0; i<_NUM_CARTAS; i++){
		sort = (int)(intptr_t)call(&this, sorteia, _NUM_CARTAS -1);
		aux = ((Object*)get(deck))[sort];
		((Object*)get(deck))[sort] = ((Object*)get(deck))[_NUM_CARTAS -i -1];
		((Object*)get(deck))[_NUM_CARTAS -i -1] = aux;
	}
	);
/*
 * Deck:imprimeDeck - 
 * Imprime o monte
 */
function(Deck, public, imprimeDeck,
	int i;
	for(i=0; i<_NUM_CARTAS; i++){
		printf("%s", (char*)call(&((Object*)get(deck))[i], Desenho));
	}
	);
/*
 * Deck:obterCarta - 
 * Retira a carta do topo do monte
 */
////// OVERRIDE de interface!!!
function(Deck, public, obterCarta,
	Object *c;
	c = &((Object*)get(deck))[(int)(intptr_t)get(topo)];
	get(topo)--;
	return (generic)c;
	);